### ISSUE TITLE
Use t.SetEnv instead of os.SetEnv

### ISSUE THREAD

**Issue by teresaromero** [okteto/okteto#3179]
Across unit test we find sometimes env vaariables are being set and then defer to restore original value. Using `t.SetEnv` performs the action of set the env just for the test and then do the cleanup.

**Comment by pchico83** [comment]
@teresaromero  could we do this in okteto/app tests to improve their isolation and allow parallel execution?

**Comment by AgustinRamiroDiaz** [comment]
> @teresaromero could we do this in okteto/app tests to improve their isolation and allow parallel execution?

@pchico83 `t.SetEnv` doesn't isolate the test, it's only a handy function that automatically cleans up the envvars at the end of your test even when it fails. To be able to isolate our tests and be able to run them in parallel, we'd need to inject in some form the `setEnv` function which will allow us to mock it. This is a trade off since we'd be reducing the scope of the test to be more `unit`, but will abstract from external dependencies (envvars)

**Comment by teresaromero** [comment]
> > @teresaromero could we do this in okteto/app tests to improve their isolation and allow parallel execution?
> 
> @pchico83 `t.SetEnv` doesn't isolate the test, it's only a handy function that automatically cleans up the envvars at the end of your test even when it fails. To be able to isolate our tests and be able to run them in parallel, we'd need to inject in some form the `setEnv` function which will allow us to mock it. This is a trade off since we'd be reducing the scope of the test to be more `unit`, but will abstract from external dependencies (envvars)

@AgustinRamiroDiaz should i edit the issue and make this about injecting envs insead?

**Comment by omnomagonz** [comment]
I created an internal backlog item [here](https://www.notion.so/okteto/Use-t-SetEnv-instead-of-os-SetEnv-f54cc10f5d9741b7b08594e0049071c7) primarily to flex the muscle of managing things in a single place.

@jmacelroy this is probably something that would skip straight to the end of `Build Track` right?

**Comment by miniscruff** [comment]
> > @teresaromero could we do this in okteto/app tests to improve their isolation and allow parallel execution?
> 
> @pchico83 `t.SetEnv` doesn't isolate the test, it's only a handy function that automatically cleans up the envvars at the end of your test even when it fails. To be able to isolate our tests and be able to run them in parallel, we'd need to inject in some form the `setEnv` function which will allow us to mock it. This is a trade off since we'd be reducing the scope of the test to be more `unit`, but will abstract from external dependencies (envvars)

I think a different approach would be to isolate the env var loading in sort of a pre-pass step, testing that using t.Setenv, then using a built struct for all other uses. This removes the need for mocking the setenv work and separates the env var loading from the actual work being done.

Something like:
```go
type ContextEnvVars struct {
    Token *string // maybe use pointers to separate no value from empty value if necessary
    Namespace *string
    ...
}

func NewContextEnvVars() *ContextEnvVars {
    var token *string
    if tokenEnv, ok := os.LookupEnv(model.OktetoTokenEnvVar); ok {
        token = *tokenEnv
    }
    return &ContextEnvVars{
        Token: token,
        ....
    }
}

// in our test file
func TestNewContextEnvVarsWithValues(t *testing.T) {
t.Setenv(model...., "Some-Token")
envStruct := NewContextEnvVars()
require.Equals(t, "Some-Token", envStruct.Token)

// then in any func that needs the env vars, just pass the env struct in as a dependency
func (c *ContextCommand) Run(ctx context.Context..., envVars *ContextEnvVars) error {
    ...
}
```

Could maybe combine it with the work related to this task: https://github.com/okteto/okteto/issues/3186 in separating the values being loaded from the env, cli args and general command values. Not sure how best to do that but seems at least partially related as env+cli args could be tested together for priority values.

**Comment by jLopezbarb** [comment]
Hi @miniscruff, We are thinking about creating a module like [afero](https://github.com/spf13/afero) but for environment variables. We thought about creating an interface with the following functions
```go
type Environment interface {
  Clearenv()
  Environ() []string
  ExpandEnv(key string) string
  GetEnv(key string) string
  LookUpEnv(key string) (bool, error)
  Setenv(key string, value string) error
  UnsetEnv(key string) error
}
```
Once the interface is created, we can implement it with the methods of the `os` library or using a `map[string]string` to store the variables. WDYT?

**Comment by miniscruff** [comment]
@jLopezbarb hmm, if I understand the idea here, it is to create a mockable environment. One that could be the real OS envs or a mock, similar to afero. The idea being with this, is we could pass this environment into our Run commands similar to what I presented and it would use that instead of the os.* funcs. 

That would solve the issue of how to parallel test work that requires env vars, but I don't think it solves the readability argument of isolating env vars from the work itself. There would be no reason you couldn't, and wouldn't just pass the environment value all throughout your call stack reading and parsing values whenever you needed to. This would make it much harder to track down what env vars are loaded and where. Issues or bugs of "under these specific combinations of env vars something goes wrong", would need to follow lots of code, where as by building a struct we could just see what the values become under those conditions and write a test.

In addition to that, your tests would sort of have this magic knowledge of exactly what env vars are used, there exact type and parsing requirements ( cause env vars are just strings but we could expect bools or ints ) and under what conditions they may or may not be read. This means tests would look like:
```go
func TestThingWithEnvVars(t *testing.T) {
    env := MockEnvironment(map[string]string{
        "MAGIC_KEY": "magic value",
        ....
    })
    err := ctxCmd.Run(..., env)
    require.Succeed(err)
}
```

To me, this goes against the static typing we sort of expect to deal with. Where if we had a typed struct, we could also include docs, util funcs or whatever specific to that grouping of env vars.

Now, there is the argument that we do have to have this knowledge of env vars somewhere, but if it was isolated to a single New*EnvVars it would be contained.

Hopefully that all made sense.
TL;DR, I would push the env vars to the boundaries of work rather then having it come along for the ride, effectively.

**Comment by miniscruff** [comment]
Thinking about it a little more, you might be able to create one struct for the environment rather than one per command or group. You would only ever use a subset of the values but they would all need to be valid at some point. And it would never be a performance issue to load a few extras imo.

**Comment by AgustinRamiroDiaz** [comment]
@jLopezbarb regarding the `Environment` struct, I think we should be careful when to use it. I think it has it's place as an intermediary in the almost top level cobra commands, before passing down the arguments to the most internal layers

But in general the abstractions defined by dependencies should be thought from the perspective of the function/method. What I mean is that we should prefer
```
func doInOktetoNamespace(oktetoNamespace string) {
  ...
}
```
rather than
```
func doInOktetoNamespace(e Environment) {
  oktetoNamespace := e.GetEnv("OKTETO_NAMESPACE")
  ...
}
```

This is a small easy example, but we should try to think the same regarding all external things like a "store" of envvars (could be map[string]string or an interface that `Environment` would satisfy) or the ports used by the OS

Just a general note to keep in mind. :D
